Added working gearshift and clutch

JohnButlerThird opened the pull request:
- Need to add model


I will add more content here, I'm new at java so please tell me if the code is adequate for this mod. I ported gearshift/clutch mostly 1:1, they use a special class for detecting redstone signals, everything seems to  work.

JohnButlerThird commented:
The Gearshift:
https://github.com/user-attachments/assets/90c1367a-c349-4144-9741-8639ea20f9df

The Clutch:

https://github.com/user-attachments/assets/04bb6114-c91c-48c3-858d-83e264f7d589

Gordon-Frohman commented:
I have to ask first: what version you are backporting from?
For the first version of ReCreate I chose version 0.4.1, the last version before the Full Steam update
Backporting stuff from newer versions would just add us more job we don't need to do right now

JohnButlerThird commented:
Added working gearbox
It synchronize with the right rotation, the placement logic and the box doesnt work but I will fix it.
https://github.com/user-attachments/assets/f0a2484a-eb22-4482-8a17-9f855f247c43

JohnButlerThird commented:
https://github.com/user-attachments/assets/eaf715c8-f398-4c92-beeb-47353bcb7650

Gordon-Frohman commented:
Oh, and one more question: are you backporting goggles tooltips as well?

JohnButlerThird commented:
> Oh, and one more question: are you backporting goggles tooltips as well?

Like which one ? I didn’t see any goggle tooltip in any block code I ported.
But once I get to consumers I will

JohnButlerThird commented:
> Great job! But you might want to pause for some time: new version of MetaWorlds will require moving some classes from ReCreate into it, so some code will have to be rewritten

The code i added is not related to meta worlds contraptions stuff simply because as you said the contraption system depends on meta worlds.

JohnButlerThird commented:
Also here is the waterwheel I ported yesterday with mostly 1:1 consistency with create 0.4.1

https://github.com/user-attachments/assets/5a9a2827-84fe-45a2-8441-653efcc010ad

Gordon-Frohman commented:
> can you pull the PR right now

I have to check it first

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/components/waterwheel/WaterWheelModel.java:
```diff
@@ -0,0 +1,132 @@
+
+    protected ModelRenderer core;
+    protected Direction.Axis axis;
+    private final ModelRenderer InnerSegment_r1;
```
Please avoid starting names of your variables with capital letters

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/relays/elementary/cogwheel/CogWheelBlock.java:
```diff
@@ -36,7 +36,7 @@ protected CogWheelBlock(Material materialIn, boolean large) {
         this.setResistance(5.0F);
         this.setStepSound(soundTypeWood);
-        this.setBlockTextureName("planks_oak");
+        this.setBlockTextureName("planks_spruce");
```
I find oak particles more fitting here. Spruce are too dark for this texture

JohnButlerThird replied:
Changed that

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/relays/encased/AbstractEncasedShaftBlock.java:
Again, half of methods missing. Keep in mind that some methods were renamed between versions

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/relays/encased/AbstractRedstoneShaftBlock.java:
Where did this class come from?

JohnButlerThird replied:
To detect redstone signal, I cannot create a POWERED blockstate, so i use this class to detect power changes, also this is what detect when its "powered".

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/relays/encased/SplitShaftModel.java:
Overcomplicated
Just use two `ShaftModel(AxisDirection direction)` instead

Gordon-Frohman commented on src/main/resources/assets/recreate/textures/models/clutch.png:
Most of your textures are too large
Canvas doesn't have to be square
Also textures themselves might have to be redone

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/relays/encased/ClutchBlock.java:
Shouldn't it extend `GearshiftBlock`?

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/RotationPropagator.java:
```diff
@@ -128,17 +134,23 @@ private static boolean isLargeToLargeGear(KineticTileEntity from, KineticTileEnt
     }
 
     private static float getAxisModifier(KineticTileEntity te, Direction direction) {
-        // TODO
```
Why did you have to rewrite all of this?

Gordon-Frohman commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/relays/encased/GearshiftBlock.java:
Once again, half of methods missing

Gordon-Frohman reviewed (changes requested):
Well... that's a mess. Hell lot of stuff to be redone, and would prefer doing it myself

JohnButlerThird commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/wrench/IWrenchable.java:
```diff
@@ -36,13 +35,12 @@ public default boolean onWrenched(World world, int x, int y, int z, int face, En
     }
 
     public default boolean onSneakWrenched(World world, int x, int y, int z, int face, EntityPlayer player) {
-        if (world instanceof WorldServer) {
```
Because this causes the wrench to give ghost items to the client. This allows the wrench to give the item with no sync issue.

JohnButlerThird commented on src/main/java/su/sergiusonesimus/recreate/compat/nei/CreateNEI.java:
I removed it on my environment.

JohnButlerThird commented on src/main/java/su/sergiusonesimus/recreate/content/contraptions/components/waterwheel/WaterWheelBlock.java:
Because I moved them to the tile entity, it seems to work the same, it is finished as it is consistent with base create.

JohnButlerThird commented:
I'm sorry of all the problems you have with the code I will try to explain a bit of what I did
- AbstractRedstoneShaftBlock is a madeup class I made to make redstone control possible because in base create they seem to add blockstate definition for powered, the redstone class handles all of the redstone stuff.

As a result I moved most of the gearshift functions there (rotationAdded etc..) because I am not able to extends Clutch with Gearshift because they both use different tile entities and this mod doesnt seem to support <? extends SplitShaftTileEntity>
So this keeps mostly every behavior 1:1, except that if you place a gearshift near a redstone powered on and its connected to a network, you need to do a block change for it to update, that's the only "problem" I've encountered..

Now for Rotation Progatator, I needed to add the same code for the gearshift then the one for the gearbox, this is not in NOT in base 0.4.1 and I acknowledge that, but the code that is is the 1.18 work doesnt seem to work here for the gearshift, as the gearshift doesn't seem to invert the rotation, since it invert the same way as a gearbox, copying the code seems to make it able to turn rotation. If I could get help onto how base create inverts the rotation.

I also needed to change the code for rendering of split shaft to synchronize them, base create 0.4.1 uses shaft instances, and I needed to make my own code as the rendering is vasly different, but my "alternative" seems to work for this case here.

I updated the waterwheel code so it is mostly moved to the block class, the WaterWheel should be pretty much the same as create 0.4.1.

After testing I think I was able to port the "waterwheel", "clutch" and "gearshift" with pretty okay consistency.
For the textures I will update them, the waterwheel one was autogenerated by blockbench.

I really hope you can understand that I tried my best, and not simply reject this PR, I'd like to contribute to the port.
I will fix the rest of the things you told me. Thanks for listening.

Gordon-Frohman commented:
Don't worry, I won't reject it
I respect all the work you've done
But Occam said "Entia non sunt multiplicanda praeter necessitatem" - "Entities must not be multiplied beyond necessity". And he wasn't talking about game entities :D
I strongly belive that same result could be achieved without adding extra classes
Also moving methods from blocks to tile entities will most likely cause some conflicts. Since I'm mostly adapting code from 0.4.1, modifying bare minimum for it to work, other classes will look for the methods you are removing in block class, not in tile entity one.

JohnButlerThird commented:
> Don't worry, I won't reject it I respect all the work you've done But Occam said "Entia non sunt multiplicanda praeter necessitatem" - "Entities must not be multiplied beyond necessity". And he wasn't talking about game entities :D I strongly belive that same result could be achieved without adding extra classes Also moving methods from blocks to tile entities will most likely cause some conflicts. Since I'm mostly adapting code from 0.4.1, modifying bare minimum for it to work, other classes will look for the methods you are removing in block class, not in tile entity one.

I moved the classes from tile to waterwheel block. 
the only thing I have yet to fix is the rendering and extra code needed for the gearshift to invert rotation.
But yeah, it’s just that trying to read the 1.18 rendering code is very hard as most of the function for rotating stuff is not present in this fork. Not sure if you are planning to port the rendering to fit new create, it seems like a lot of work.
What I did works but I needed to write a lot of my own rendering code for it to work the same as 1.18

Gordon-Frohman commented:
Rendering is one of the only things I do from scratch
It is way too different in 1.7.10

JohnButlerThird commented:
> Don't worry, I won't reject it I respect all the work you've done But Occam said "Entia non sunt multiplicanda praeter necessitatem" - "Entities must not be multiplied beyond necessity". And he wasn't talking about game entities :D I strongly belive that same result could be achieved without adding extra classes Also moving methods from blocks to tile entities will most likely cause some conflicts. Since I'm mostly adapting code from 0.4.1, modifying bare minimum for it to work, other classes will look for the methods you are removing in block class, not in tile entity one.

Would you rather I move the function of abstract red stone to be straight into gearshift ? And also how can I make it so I can extends the gearshift and not error when trying to create a clutch tile entity extending gearshift.
If you can tell me how, I will delete my extra class and make it like 1.18.

JohnButlerThird commented:
> Rendering is one of the only things I do from scratch It is way too different in 1.7.10

I was able to copy rendering code from modern and adapt it for gearbox and gearshift. You can try in creative I think it’s the best I could’ve done.

Gordon-Frohman commented:
> Would you rather I move the function of abstract red stone to be straight into gearshift ? And also how can I make it so I can extends the gearshift and not error when trying to create a clutch tile entity extending gearshift.
> If you can tell me how, I will delete my extra class and make it like 1.18.

For now I'm not sure. I need to check both 1.7.10 and 1.18 sources to decide

JohnButlerThird commented:
> > Would you rather I move the function of abstract red stone to be straight into gearshift ? And also how can I make it so I can extends the gearshift and not error when trying to create a clutch tile entity extending gearshift.
> > If you can tell me how, I will delete my extra class and make it like 1.18.
> 
> For now I'm not sure. I need to check both 1.7.10 and 1.18 sources to decide

1.18 has blockentitytype so the gearshift has split shafttileentity but has gearshifttileentity as the type.
Doesn’t seem to exist in 1.7, so I can’t make the clutch extends the gearshift.

JohnButlerThird commented:
From what I modified what are the things you think I should modify before being PR^
I told you that for the rendering/logic of gearshift and clutch, I was only able to do it this way.
For textures, I guess I need to remake them.

JohnButlerThird commented:
I gave full permission to my repo.

Gordon-Frohman commented:
Hmm, I'm still unable to push
Is editing of this PR by the contributors enabled?

Gordon-Frohman commented:
I've also made some changes in gearbox-related classes, but before I add all those, I need to try something render-related. There might be a nicer way to render shafts and other blocks

JohnButlerThird commented:
Yes the problem is that the mod is currently renderer with TESR, we should try to renter static parts in the simple block renderer.
Or maybe even draw everything instead of the chunk mesh, I heard it's even better instead of TESR.

Also for the gearshift, god hell i needed to substract the coordinates, I completely missed that, thanks for cleaning that up.

Gordon-Frohman commented:
I was thinking about switching to just using RenderBlocks for static parts
But we can at least use it for creative motor, I guess

JohnButlerThird commented:
> I was thinking about switching to just using RenderBlocks for static parts But we can at least use it for creative motor, I guess

Well this is what it does, it’s just that you can keep your models in model base and then just render them with render blocks

Gordon-Frohman commented:
Well, now it is done
I'm very grateful for your contribution, especially for the FluidHelper class, but if there's one thing I learned from working on this PR, it's the fact that I hate fixing stuff someone else made instead of making it myself
Therefore, from now on I will keep ReCreate as a solo project
Once again, thank you for your work

JohnButlerThird commented:
Wouldn't it be better that you just tell me what standard you want me to adopt, so you dont need to fix it yourself?

JohnButlerThird commented:
I don't think I'll be modding for a big while, I just ask for the future if I ever want to contribute, what standard do you expect.

Gordon-Frohman commented:
Exact copy of the original Create code, with minor changes to make it work on 1.7.10.
And if you are making models, you don't have to copy same texture for every repeating part, you can just make every model point to the same area of the texture.
You can check all the changes that I made to your code and look up to them.
